ocean_lib/common/ddo/credentials.py

```diff
@@ -0,0 +1,106 @@
+    def __init__(self, asset):
+        self.asset = asset
+
+    def get_addresses_of_class(self, access_class="allow"):
+        """Get a filtered list of addresses from credentials (use with allow/deny)."""
+        entries = self.asset._credentials.get(access_class, [])
+
+        for entry in entries:
+            if entry["type"] == "address":
+                return [val.lower() for val in entry["values"]]
+
+        return []
+
+    def requires_credential(self):
+        allowed_addresses = self.get_addresses_of_class("allow")
+        denied_addresses = self.get_addresses_of_class("deny")
+
+        return allowed_addresses or denied_addresses
+
+    def validate_access(self, credential=None):
+        address = simplify_credential_to_address(credential)
+
+        allowed_addresses = self.get_addresses_of_class("allow")
+        denied_addresses = self.get_addresses_of_class("deny")
+
+        if not address and not self.requires_credential():
+            return ConsumableCodes.OK
+
+        if allowed_addresses and address.lower() not in allowed_addresses:
+            return ConsumableCodes.CREDENTIAL_NOT_IN_ALLOW_LIST
+
+        if not allowed_addresses and address.lower() in denied_addresses:
+            return ConsumableCodes.CREDENTIAL_IN_DENY_LIST
+
+        return ConsumableCodes.OK
+
+    def add_address_to_list_class(self, address, list_class="allow"):
+        address = address.lower()
+
+        if not self.asset._credentials or list_class not in self.asset._credentials:
+            self.asset._credentials[list_class] = [
+                {"type": "address", "values": [address]}
+            ]
+            return
+
+        address_entry = get_list_entry_with_type_address(
+            self.asset._credentials[list_class]
+        )
+
+        if not address_entry:
+            self.asset._credentials[list_class].append(
+                {"type": "address", "values": [address]}
+            )
+            return
+
+        address_entry = address_entry[0]
+        lc_addresses = [addr.lower() for addr in address_entry["values"]]
+
+        if address not in lc_addresses:
+            lc_addresses.append(address)
+
+        address_entry["values"] = lc_addresses
+
+    def remove_address_from_list_class(self, address, list_class="allow"):
+        address = address.lower()
+
+        if not self.asset._credentials or list_class not in self.asset._credentials:
+            return
+
+        address_entry = get_list_entry_with_type_address(
+            self.asset._credentials[list_class]
+        )
+
+        if not address_entry:
+            return
+
+        address_entry = address_entry[0]
+        lc_addresses = [addr.lower() for addr in address_entry["values"]]
+
+        if address not in lc_addresses:
+            return
+
+        lc_addresses.remove(address)
+        address_entry["values"] = lc_addresses
+
+
+def simplify_credential_to_address(credential):
+    if not credential:
+        return None
```

Comment:
Shouldn't a MalformedCredential error be raised at this point as well?

Reply:
no "Credential" means no dictionary as well, i.e. there is no address credential, nor other type of credential. In the future we might chain multiple credentials, so I check for the "requires address" separately.

Reply:
I don't understand your last comment. It sounds like you're talking about the `credentials` (plural) data structure described by https://github.com/oceanprotocol/multi-repo-issue/issues/98.

The `simplify_credential_to_address` function extracts the "value" from a `credential` (singular) which looks like `{"type": "something", "value": "something"}`.  If the caller passes anything that doesn't match that schema, this function should raise a `MalformedCredential` error.

Reply:
The credential can be empty and the address can be None, it still needs checking. E.g. explicitly or implicitly passing empty. We still have to check if they are allowed. The case is covered by the same functions for allow and deny. 

Maybe I can't "see" these anymore, since I've spun them around too much. I'll look into it this morning. I'll submit answers for the other two comments.